lib/vrl/compiler/src/expression/function_closure.rs

```diff
@@ -0,0 +1,194 @@
+                    state.insert_variable(value_ident.clone(), value);
+
+                    let output = self.block.resolve(ctx)?;
+
+                    result = func(ctx, output, result)?;
+
+                    let state = ctx.state_mut();
+                    state.remove_variable(&key_ident);
+                    state.remove_variable(&value_ident);
+                }
+                Ok(result)
+            }
+            Value::Array(array) => {
+                let index_ident = self
+                    .variables
+                    .get(0)
+                    .expect("checked at compile-time")
+                    .clone()
+                    .into_inner();
+                let value_ident = self
+                    .variables
+                    .get(1)
+                    .expect("checked at compile-time")
+                    .clone()
+                    .into_inner();
+
+                for (index, value) in ArrayRecursive::create_iter(array, recursive) {
+                    let state = ctx.state_mut();
+                    state.insert_variable(index_ident.clone(), index.into());
+                    state.insert_variable(value_ident.clone(), value);
+
+                    let output = self.block.resolve(ctx)?;
+
+                    result = func(ctx, output, result)?;
+
+                    let state = ctx.state_mut();
+                    state.remove_variable(&index_ident);
+                    state.remove_variable(&value_ident);
+                }
+                Ok(result)
+            }
+            _ => unimplemented!(),
+        }
+    }
+}
+
+impl fmt::Display for FunctionClosure {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // TODO
+        f.write_str("{ |")?;
+
+        let mut iter = self.variables.iter().peekable();
+        while let Some(var) = iter.next() {
+            var.fmt(f)?;
+
+            if iter.peek().is_some() {
+                f.write_str(", ")?;
+            }
+        }
+
+        f.write_str("|\n")?;
+        self.block.fmt(f)?;
+
+        f.write_str("\n}")
+    }
+}
+
+struct ObjectRecursive {
+    queue: VecDeque<(String, Value)>,
+}
+
+impl ObjectRecursive {
+    fn new(map: BTreeMap<String, Value>) -> Self {
+        let queue = map.into_iter().collect();
+        Self { queue }
+    }
+
+    fn create_iter(
+        map: BTreeMap<String, Value>,
+        recursive: bool,
+    ) -> Box<dyn Iterator<Item = (String, Value)>> {
+        if recursive {
+            Box::new(ObjectRecursive::new(map).into_iter())
+        } else {
+            Box::new(map.into_iter())
+        }
+    }
+}
+
+impl Iterator for ObjectRecursive {
+    type Item = (String, Value);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (key, value) = self.queue.pop_front()?;
+        match value {
+            Value::Object(ref map) => {
+                self.queue.extend(map.clone().into_iter());
```

Comment:
- Are we comfortable with the amount of cloning that would happen for deeply nested values in this current code?
- Is it a reasonable default to have the last encountered key-value pair be the value that remains in the object in the event of a key collision? e.g. `map({"a": 1, "b": {"a":2}}, true) -> |key, value| { [key, value]} == {"a":2, "b": {"a":2}}`